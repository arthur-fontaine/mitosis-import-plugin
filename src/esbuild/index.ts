import type { Plugin } from 'esbuild';
import { mitosisImportPluginCore } from '../core/mitosisImportPluginCore';

export const mitosisImportPlugin = (): Plugin => ({

  name: 'mitosis-import',

  setup(build) {

    build.onResolve({ filter: /.*/ }, (args) => {

      if (!mitosisImportPluginCore.requiresMitosisProcessing(args.with)) return null;

      const target = mitosisImportPluginCore.getMitosisTarget(args.with);

      return {
        namespace: 'mitosis',
        pluginData: { target },
      };

    });

    build.onLoad({ filter: /.*/, namespace: 'mitosis' }, async (args) => {

      const target = args.pluginData?.target;
      if (!target) throw new Error('Target not found in plugin data');

      const compiledComponent = await mitosisImportPluginCore.compileMitosisComponent(args.path, args.namespace, target);

      return {
        contents: compiledComponent,
        loader: 'tsx' as const, // TODO: can esbuild be compatible with vue?
      };

    })

  },

});
