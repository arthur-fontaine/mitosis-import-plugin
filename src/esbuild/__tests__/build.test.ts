import esbuild from "esbuild";
import { mitosisImportPlugin } from "..";
import { describe, expect, test } from "vitest";

describe("esbuild build", () => {
	test("should build with mitosis import plugin", async () => {
		const result = await esbuild.build({
			stdin: {
				contents: `
          import MyComponent from "./component" with { mitosis: "react" };
          export const comp = () => <MyComponent />;
        `,
				loader: "jsx",
				resolveDir: __dirname,
			},
			write: false,
			plugins: [mitosisImportPlugin()],
			bundle: true,
			external: ["react"],
			format: "esm",
		});

		const [outputFile] = result.outputFiles;

		expect(outputFile.text).toBeTruthy();
		expect(outputFile.text).not.toContain('with { mitosis: "react" }');
		expect(outputFile.text).toMatchSnapshot();
	});

	test("should build with mitosis import plugin and forced target", async () => {
		const result = await esbuild.build({
			stdin: {
				contents: `
					import MyComponent from "./component" with { mitosis: "BAD TARGET" };
					export const comp = () => <MyComponent />;
				`,
				loader: "jsx",
				resolveDir: __dirname,
			},
			write: false,
			plugins: [mitosisImportPlugin({ target: "react" })],
			bundle: true,
			external: ["react"],
			format: "esm",
		});

		const [outputFile] = result.outputFiles;

		expect(outputFile.text).toBeTruthy();
		expect(outputFile.text).not.toContain('with { mitosis: "react" }');
		expect(outputFile.text).toMatchSnapshot();
	});

	test("should detect mitosis files with source and build", async () => {
		const result = await esbuild.build({
			stdin: {
				contents: `
					import MyComponent from "./component";
					export const comp = () => <MyComponent />;
				`,
				loader: "jsx",
				resolveDir: __dirname,
			},
			write: false,
			plugins: [mitosisImportPlugin({
				detectMitosisFilesWithSource: true,
				target: "react",
			})],
			bundle: true,
			external: ["react"],
			format: "esm",
		});

		const [outputFile] = result.outputFiles;

		expect(outputFile.text).toBeTruthy();
		expect(outputFile.text).toMatchSnapshot();
	});
});
