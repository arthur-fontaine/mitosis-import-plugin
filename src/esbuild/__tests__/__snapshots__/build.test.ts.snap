// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`esbuild build > should build with mitosis import plugin 1`] = `
"// mitosis:./component
import * as React2 from "react";
function MyComponent(props) {
  const state = useLocalProxy({ name: "Steve" });
  return /* @__PURE__ */ React2.createElement("div", null, /* @__PURE__ */ React2.createElement(
    "input",
    {
      css: {
        color: "red"
      },
      value: state.name,
      onChange: (event) => state.name = event.target.value
    }
  ), "Hello! I can run natively in React, Vue, Svelte, Qwik, and many more frameworks!");
}
var component_default = MyComponent;

// <stdin>
var comp = () => /* @__PURE__ */ React.createElement(component_default, null);
export {
  comp
};
"
`;
